Add StudioCMS guide and assets to CMS Section (#11285)
withastro/docs: src/content/docs/en/guides/cms/studiocms.mdx

---
title: StudioCMS & Astro
description: Build and manage content for your Astro project using StudioCMS, a headless CMS designed specifically for Astro.
sidebar:
  label: StudioCMS
type: cms
service: StudioCMS

> stub: false

Reviewer: ```suggestion
```

---
title: StudioCMS & Astro
description: Build and manage content for your Astro project using StudioCMS, a headless CMS designed specifically for Astro.
sidebar:
  label: StudioCMS
type: cms
service: StudioCMS
stub: false
i18nReady: true
---
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'
import { FileTree } from '@astrojs/starlight/components';
import { Steps } from '@astrojs/starlight/components';

> [StudioCMS](https://studiocms.dev/) is a headless CMS designed specifically for Astro that provides a user-friendly dashboard for content management with zero client-side Frameworks.

Author: > > `[StudioCMS](https://studiocms.dev/) is a Astro native`
> 
> Could we add an `n` here? `[StudioCMS](https://studiocms.dev/) is an Astro native`
> 
> Also, I think the comma behind `offering` is meant to go behind `management` instead!

updated

> [StudioCMS](https://studiocms.dev/) is a headless CMS designed specifically for Astro that provides a user-friendly dashboard for content management with zero client-side Frameworks.

Author: @sarah11918 your input here? :sweat_smile:

> [StudioCMS](https://studiocms.dev/) is a headless CMS designed specifically for Astro that provides a user-friendly dashboard for content management with zero client-side Frameworks.

Reviewer: ```suggestion
[StudioCMS](https://studiocms.dev/) is a headless CMS for Astro, built with Astro, that provides a user-friendly and configurable dashboard for content management as well as a custom rendering system to display your Astro components.
```

     pnpm astro db push --remote
     ```
     </Fragment>
     <Fragment slot="yarn">
     ```bash
     yarn astro db push --remote
     ```
     </Fragment>
   </PackageManagerTabs>

2. Start your development server:
   <PackageManagerTabs>
     <Fragment slot="npm">
     ```bash
     npm run dev
     ```
     </Fragment>
     <Fragment slot="pnpm">
     ```bash
     pnpm run dev
     ```
     </Fragment>
     <Fragment slot="yarn">
     ```bash
     yarn dev
     ```
     </Fragment>
   </PackageManagerTabs>

3. Complete the setup by visiting: `http://localhost:4321/start`
</Steps>

Your StudioCMS dashboard will be available at `http://localhost:4321/dashboard` in development and `https://your-domain.tld/dashboard` in production.

## Adding a frontend to your StudioCMS project

StudioCMS is headless, so you'll need to build your own frontend or use a plugin. To add the blog plugin:

<PackageManagerTabs>
  <Fragment slot="npm">
  ```bash
  npm install @studiocms/blog
  ```
  </Fragment>
  <Fragment slot="pnpm">
  ```bash
  pnpm add @studiocms/blog
  ```
  </Fragment>
  <Fragment slot="yarn">
  ```bash
  yarn add @studiocms/blog
  ```
  </Fragment>
</PackageManagerTabs>

> Then configure it in your `studiocms.config.mjs` file:

Reviewer 2: This is the first mention I saw of a `studiocms.config.mjs` file, so perhaps this should instead say something like

"Then create a `studiocms.config.mjs` file with the following structure in the root of your project" (or wherever it should be)

Unless this file is generated on install in which case that should probably be mentioned on lines [97-102](https://github.com/withastro/docs/pull/11285/files#diff-fac55326c66fa02e69f0149bb8bc267ab1849d343c6fc2e68a981ec20fdcb498R97-R102)

Author: just pushed an update :sweat_smile: small oversight from multiple people doing this

---

> title: StudioCMS

Reviewer 3: A last nit to be consistent with other guides (e.g., [Storyblok](https://docs.astro.build/en/guides/cms/storyblok/)):
```suggestion
title: StudioCMS & Astro
```

import { FileTree } from '@astrojs/starlight/components';
import { Steps } from '@astrojs/starlight/components';

[StudioCMS](https://studiocms.dev/) is a headless CMS designed specifically for Astro that provides a user-friendly dashboard for content management with zero client-side JavaScript.

This guide will show you how to set up StudioCMS in your Astro project, configure authentication, and start managing your content.

## Prerequisites

To get started, you will need to have the following:

* **An Astro project** - You can use an existing Astro project or create a new one.
* **A libSQL provider or self-hosted libSQL server** - StudioCMS uses [`@astrojs/db`](/en/guides/astro-db/) to connect to your database.
* **An adapter that supports SSR** - StudioCMS requires an [adapter that supports SSR Routes](/en/guides/on-demand-rendering/#server-adapters).

## Setting up your database

StudioCMS works with any libSQL provider. [Turso](https://tur.so/studiocms/) is recommended and offers a discount with the code `STUDIOCMS10` for 10% off for 12 months.

<Steps>
  1. Install and set up a libSQL provider such as Turso:
      ```bash

>       curl -sSfL https://get.turso.tech/install.sh | bash

Reviewer 3: It seems the link was wrong. With `turso.tech` I got "Could not resolve host". In [Turso docs](https://docs.turso.tech/cli/installation#linux) I see:
```suggestion
      curl -sSfL https://get.tur.so/install.sh | bash
```

>       turso auth login

Reviewer 3: Is this step necessary? With the previous command, Turso wants me to log in in the browser. So running this seems useless: "Already signed in as...".
But it's not necessarily annoying, it was just to note it down, just in case. It's better to be sure you're connected. 😄

Author: It shouldn't be needed... but it could be depending on the system, or if something doesn't go quite as planned during install

Reviewer 3: Makes sense, so might as well leave it! I'll mark this as resolved.

        ```
        </Fragment>
        <Fragment slot="yarn">
        ```bash
        yarn add @astrojs/db @astrojs/node studiocms
        ```
        </Fragment>
      </PackageManagerTabs>

  2. Update your `astro.config.mjs` file:

      ```js title="astro.config.mjs"
      import { defineConfig } from 'astro/config';
      import db from '@astrojs/db';
      import node from '@astrojs/node';
      import studioCMS from 'studiocms';

      export default defineConfig({
          site: 'https://your-site.com/', // Required for StudioCMS
          output: 'server',
          adapter: node({ mode: "standalone" }),
          integrations: [
              db(),
              studioCMS(),
          ],
      });
      ```

  3. (Optional) Create a `studiocms.config.mjs` file to configure StudioCMS:

      ```js title="studiocms.config.mjs"
      import { defineStudioCMSConfig } from 'studiocms/config';

      export default defineStudioCMSConfig({
          dbStartPage: true, // You will set this to false in a later step
      });
      ```
</Steps>

Your project should now include these files:

<FileTree title="Project Structure">
- src/
- .env
- astro.config.mjs

> - studiocms.config.mjs (optional if your not using plugins)

Reviewer 4: ```suggestion
- studiocms.config.mjs (optional if you're not using plugins)
```

- package.json
</FileTree>

## Configuring authentication

StudioCMS requires an encryption key for authentication. You can generate one using OpenSSL:

```bash
openssl rand --base64 16
```

Add the key to your `.env` file:

```ini title=".env"
# Required for StudioCMS authentication
CMS_ENCRYPTION_KEY="wqR+w...sRcg=="
```

### Optional: Setting up OAuth providers

StudioCMS supports authentication with GitHub, Discord, Google, and Auth0. To configure:

1. Set up OAuth applications with your chosen providers
2. Use the callback URL format: `https://your-domain.tld/studiocms_api/auth/<provider>/callback`
3. Add the credentials to your `.env` file

## Setting up package.json scripts

Update your `package.json` to include the necessary scripts:

> ```json title="package.json"

Reviewer 3: Just a nit: maybe we could highlight the part that is required to make it more obvious.
````suggestion
```json title="package.json" "--remote"
````
Especially since a new Astro installation will add `"preview": "astro preview"` in the scripts so `--remote` will be required too, right? With the highlighting this gives a hint to the user I think.

{
    "scripts": {
        "dev": "astro dev --remote",
        "build": "astro check & astro build --remote",
        "astro": "astro"
    }
}
```

## Running your StudioCMS project

<Steps>
1. Initialize your database schema:
   <PackageManagerTabs>
     <Fragment slot="npm">
     ```bash
     npx astro db push --remote
     ```
     </Fragment>
     <Fragment slot="pnpm">
     ```bash
     pnpm astro db push --remote
     ```
     </Fragment>
     <Fragment slot="yarn">
     ```bash
     yarn astro db push --remote
     ```
     </Fragment>
   </PackageManagerTabs>

2. Start your development server:
   <PackageManagerTabs>
     <Fragment slot="npm">
     ```bash
     npm run dev
     ```
     </Fragment>
     <Fragment slot="pnpm">
     ```bash
     pnpm run dev
     ```
     </Fragment>
     <Fragment slot="yarn">
     ```bash
     yarn dev
     ```
     </Fragment>
   </PackageManagerTabs>

3. Complete the setup by visiting: `http://localhost:4321/start`
</Steps>

> Your StudioCMS dashboard will be available at `http://localhost:4321/dashboard` in development and `https://your-domain.tld/dashboard` in production.

Reviewer: ```suggestion
Your StudioCMS dashboard will be available at `http://localhost:4321/dashboard` in development and `https://example.com/dashboard` in production.
```

## Adding a frontend to your StudioCMS project

> StudioCMS is headless, so you'll need to build your own frontend or use a plugin. To add the blog plugin:

Reviewer: I did kind of laugh at what sounds like another step, like,  "Step 8: build your own front end!"

What about something like...

```suggestion
StudioCMS provides a [blog plugin](https://docs.studiocms.dev/en/package-catalog/studiocms-plugins/studiocms-blog/) that will create a blog site for you and allow you to create, edit, and delete blog posts from the StudioCMS dashboard.

To get started with a Studio CMS blog, first install the plugin:
```

<PackageManagerTabs>
  <Fragment slot="npm">
  ```bash
  npm install @studiocms/blog
  ```
  </Fragment>
  <Fragment slot="pnpm">
  ```bash
  pnpm add @studiocms/blog
  ```
  </Fragment>
  <Fragment slot="yarn">
  ```bash
  yarn add @studiocms/blog
  ```
  </Fragment>
</PackageManagerTabs>

Then configure it in your `studiocms.config.mjs` file:

```js title="studiocms.config.mjs"
import { defineStudioCMSConfig } from 'studiocms/config';
import blog from '@studiocms/blog';

export default defineStudioCMSConfig({

>     dbStartPage: false,

Reviewer: Expressive code can help here, but *here* is where you could choose to do something like:

```suggestion
    dbStartPage: false, // not needed with the blog plugin
```

Author: I'll have to clarify the section above where we set this for sure... it is used to enable a special mode in StudioCMS that specifically tells StudioCMS to boot up in a "setup mode" to be able to setup your tables and basic data such as SiteConfig and the owner account, once you complete this setup the page tells you to specifically set it to `false`.

    plugins: [
        blog(),
    ],
});
```

> ## Official resources
> 
> - [StudioCMS Documentation](https://docs.studiocms.dev/)
> - [Environment Variables Reference](https://docs.studiocms.dev/start-here/environment-variables/)
> - [StudioCMS Configuration Reference](https://docs.studiocms.dev/config-reference/)
> - [Package Catalog](https://docs.studiocms.dev/package-catalog/)
> 
> ## Community resources
> 
> - [StudioCMS GitHub Repository](https://github.com/studiocms/studiocms)
> - [StudioCMS Discord Community](https://discord.gg/studiocms)

Reviewer: ```suggestion
## Official resources

- [StudioCMS Documentation](https://docs.studiocms.dev/)
- [StudioCMS GitHub Repository](https://github.com/studiocms/studiocms)
- [StudioCMS Discord Community](https://discord.gg/studiocms)
```
a) Aren't these *all* yours? (official)
b) We typically wouldn't list multiple items from the same source (docs). It's sufficient to just say "Here are the docs" and that's where ALL the docs are.
